api/task_manager.cc

```diff
@@ -0,0 +1,165 @@
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.sequence_number = status.task_status.sequence_number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.progress.total;
+    res.progress_completed = status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        co_return coroutine::return_exception(httpd::bad_param_exception("Task not found"));
+    }
+    auto progress = co_await task->get_progress();
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module()->get_name();
+    s.progress.completed = progress.completed;
+    s.progress.total = progress.total;
+    co_return make_status(s);
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        std::vector<std::string> v = boost::copy_range<std::vector<std::string>>(ctx.tm.local().get_modules() | boost::adaptors::map_keys);
+        co_return v;
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        using chunked_stats = utils::chunked_vector<task_stats>;
+        std::vector<chunked_stats> res = co_await ctx.tm.map([&req] (task_manager& tm) {
+            chunked_stats local_res;
+            auto module = tm.find_module(req->param["module"]);
+            const auto& filtered_tasks = module->get_tasks() | boost::adaptors::filtered([&params = req->query_parameters] (const auto& task) {
+                return filter_tasks(task.second, params);
+            });
+            for (auto& [task_id, task] : filtered_tasks) {
+                task_stats ts;
+                ts.task_id = task->id().to_sstring();
+                ts.state = task->get_status().state;
+                local_res.push_back(ts);
+            }
+            return local_res;
+        });
+
+        std::function<future<>(output_stream<char>&&)> f = [r = std::move(res)] (output_stream<char>&& s) -> future<> {
+            bool first = true;
+            auto res = std::move(r);
+            co_await s.write("[");
+            for (auto& v: res) {
+                for (auto& stats: v) {
+                    auto fut = first ? make_ready_future<>() : s.write(", ");
+                    first = false;
+                    co_await std::move(fut);
+                    tm::task_stats ts;
+                    ts = stats;
+                    co_await formatter::write(s, ts);
```

Comment:
How about `co_await formatter::write(s, tm::task_stats{stats});`?

Reply:
Unfortunately, there is no such constructor for `tm::task_stats`. It can be done either with `operator=` or `update()` method.

Reply:
A respective part of `task_manager.json.hh`:
```
task_stats(const task_stats & e) {
  register_params();
  task_id = e.task_id;
  state = e.state;

}
template<class T>
task_stats& operator=(const T& e) {
  task_id = e.task_id;
  state = e.state;

  return *this;
}
task_stats& operator=(const task_stats& e) {
  task_id = e.task_id;
  state = e.state;

  return *this;
}
template<class T>
task_stats& update(T& e) {
  e.task_id = task_id;
  e.state = state;

  return *this;
}
```

Reply:
I see.  If `tm::task_stats ts = stats;` doesn't work either, let's keep it this way...